docs/02-use-cases/06-batch-job.md

```diff
@@ -4,10 +4,121 @@ title: Batch job
+
+## Batch jobs with heartbeating
+
+Cadence is able to coordinate, restart and track progress of large batch jobs by keeping track of their incremental progress and allowing them to resume if they're stopped for any reason. This predominantly relies on the `heartbeat` feature and activity retries. 
+
+This is used in production for customers who wish to work through large batch workloads 
+
+### Considerations before starting
+
+Heartbeating cadence workers are workers who emit their progress at an appropriate interval (usually every few seconds) indicating where they are up to. Optionally, they may use progress information (like an offset number or iterator) to resume their progress. However, this necessarily implies that:
+
+- If activities get restarted, they may redo some work, so this is not suitable for non-idempotent operations.
+- The activity will be handling all the progress, so apart from heartbeat information, debugging about the granular operations being performed is not necessarily visible as compared by doing each operation in a distinct activity. 
+
+### What problems this solves
+
+- This is for high-throughput operations where work may able to fit into a single long-running activity, or partitioned across multiple activities which can run for a longer duration.
+- This addresses problems customers may have running amounts of records through a cadence workflow where returning large blocks of data in each activity, or running large numbers of activities may hit resource limits (like the number of history entries) in a cadence workflow.
+
+### High level concept:
+
+The idea is to create an activity which will handle a lot of records and record its progress:
+
+```golang
+func (a *ABatchActivity) Execute(ctx context.Context, params Params) error {
+
+    // in this case this is just a struct with a mutex protecting it
+	var state State
+
+    // when starting the activity, check at start time for a previous iteration 
+	if activity.HasHeartbeatDetails(ctx) {
+		err := activity.GetHeartbeatDetails(ctx, &state)
+		log.Info("resuming from a previous state", zap.Any("state", state))
+		if err != nil {
+			return err
+		}
+	}
+
+    // in the background, every 5 seconds, emit where we're up to
+    // so the cadence server knows the activity is still alive, and 
+    // put the progress in the recordheartbeat call so it can be pulled in
+    // if we have to restart
+    go func() {
+        ticker := time.NewTicker(time.Seconds * 5)
+        defer ticker.Stop()
+        for {
+            select {
+            case <-ctx.Done():
+                return
+            case <-ticker.C:
+                state := state.GetLocked()
```

Comment:
```suggestion
                reportState := state.Clone()
```
^ maybe?

`GetLocked()` doesn't seem to imply to me that it was a `lock -> tmp=copy -> unlock -> return tmp`.  and if it wasn't copied, serializing it will generally bypass the lock and that just races again.